packages/blocks/src/store/selectors.js

```diff
@@ -262,17 +262,28 @@ export function getActiveBlockVariation( state, blockName, attributes, scope ) {
 				continue;
 			}
 			const isMatch = definedAttributes.every( ( attribute ) => {
-				const attributeValue = getValueFromObjectPath(
-					attributes,
+				const variationAttributeValue = getValueFromObjectPath(
+					variation.attributes,
 					attribute
 				);
-				if ( attributeValue === undefined ) {
+				if ( variationAttributeValue === undefined ) {
 					return false;
 				}
-				return (
-					attributeValue ===
-					getValueFromObjectPath( variation.attributes, attribute )
+				const blockAttributeValue = getValueFromObjectPath(
+					attributes,
+					attribute
 				);
+				// If the attribute value is an object, we need to compare its properties.
+				if (
+					variationAttributeValue !== null &&
+					typeof variationAttributeValue === 'object'
```

Comment:
In Interactivity, we also check that `variationAttributeValue.constructor === Object`.
Would should check that we need here too.

Reply:
Ah, interesting! I found a few instances of `isObject` helpers across our codebase that didn't seem to include that criterion 🤔 E.g. https://github.com/WordPress/gutenberg/blob/eaf2d6550b39587dc8945bf2e20fc4290a7139fe/packages/blocks/src/api/registration.js#L127-L129

I guess it's to exclude arrays?

Reply:
Yeah, we might want to do that (also see https://github.com/WordPress/gutenberg/pull/62272#discussion_r1627318785). I'll add the check.

Reply:
1b408c75aec